File: go/bkv.go
package bkv

import (
	"encoding/hex"
	"errors"
	"fmt"
)

var (
	ErrKVNotExists = errors.New("kv not exists")
)

type KV struct {
	isStringKey bool
	key         []byte
	value       []byte
}

func NewKVFromIntegerKey(key uint64, value []byte) *KV {
	return &KV{
		isStringKey: false,
		key:         EncodeNumber(key),
		value:       value,
	}
}

func NewKVFromStringKey(key string, value []byte) *KV {
	keyBytes := []byte(key)
	if len(keyBytes) > 128 {
		keyBytes = keyBytes[0:128]
	}
	return &KV{
		isStringKey: true,
		key:         keyBytes,
		value:       value,
	}
}

// total length bytes + Key length byte(1 bit number / string flag + 7 bit length) + Key bytes + Value bytes

func (v *KV) Pack() ([]byte, error) {
	bs := make([]byte, 0)

	keyLen := len(v.key)
	if keyLen > 128 {
		return nil, errors.New(fmt.Sprintf("key length is bigger than 128: %v", keyLen))
	}

	totalLen := 1 + keyLen + len(v.value)

	keyLenByte := byte(keyLen & 0x7F)
	if v.isStringKey {
		keyLenByte |= 0x80
	}

	bs = append(bs, EncodeLength(uint64(totalLen))...)
	bs = append(bs, keyLenByte)
	bs = append(bs, v.key...)
	bs = append(bs, v.value...)

	return bs, nil
}

func (v *KV) Clone() *KV {
	key := make([]byte, 0)
	key = append(key, v.key...)

	value := make([]byte, 0)
	value = append(value, v.value...)

	return &KV{
		isStringKey: v.isStringKey,
		key:         key,
		value:       value,
	}
}

func (v *KV) IsStringKey() bool {
	return v.isStringKey
}

func (v *KV) StringKey() string {
	return string(v.key)
}

func (v *KV) NumberKey() uint64 {
	return DecodeNumber(v.key)
}

func (v *KV) Key() interface{} {
	if v.isStringKey {
		return v.StringKey()
	}

	return v.NumberKey()
}

func (v *KV) StringValue() string {
	return string(v.value)
}

func (v *KV) NumberValue() uint64 {
	return DecodeNumber(v.value)
}

func (v *KV) Value() []byte {
	return v.value
}

var (
	ErrorEmptyBuf = errors.New("empty buf")
)

func UnpackKV(buf []byte) (*KV, []byte, error) {
	if len(buf) == 0 {
		return nil, nil, ErrorEmptyBuf
	}

	err, totalLen, _, pb := DecodeLength(buf)
	if err != nil {
		return nil, buf, errors.New(fmt.Sprintf("decode total length fail: %v", err))
	}

	if uint64(len(pb)) < totalLen || len(pb) == 0 {
		return nil, buf, errors.New(fmt.Sprintf("totalLen = %v, len(pb) = %v", totalLen, len(pb)))
	}

	keyLenByte := pb[0]
	keyLen := uint64(keyLenByte & 0x7F)
	isStringKey := false
	if keyLenByte & 0x80 != 0 {
		isStringKey = true
	}

	if keyLen + 1 > totalLen {
		return nil, buf, errors.New(fmt.Sprintf("keyLen bigger than totalLen, keyLen=%v, totalLen=%v", keyLen, totalLen))
	}

	keyBytes := pb[1:(keyLen + 1)]
	valueBytes := pb[(keyLen + 1):totalLen]

	return &KV{
		isStringKey: isStringKey,
		key:         keyBytes,
		value:       valueBytes,
	}, pb[totalLen:], nil
}



type BKV struct {
	kvs []*KV
}



func NewBKV() *BKV {
	return &BKV{
		kvs: make([]*KV, 0),
	}
}

func Unpack(buf []byte) (*BKV, []byte, error) {
	bkv := &BKV{}

	for {
		// logger.Debug("unpacking buf: %v", hex.EncodeToString(buf))
		kv, pb, err := UnpackKV(buf)
		if err != nil {
			if err == ErrorEmptyBuf {
				return bkv, nil, nil
			}
			return bkv, pb, errors.New(fmt.Sprintf("parsing kv fail: %v", err.Error()))
		}
		if kv != nil {
			bkv.AddKV(kv)
		}
		buf = pb
	}
}

func (v *BKV) Pack() ([]byte, error) {
	bs := make([]byte, 0)

	for _, kv := range v.kvs {
		buf, err := kv.Pack()
		if err != nil {
			return nil, err
		}
		bs = append(bs, buf...)
	}

	return bs, nil
}

func (v *BKV) Items() []*KV {
	items := make([]*KV, 0)
	for _, item := range v.kvs {
		items = append(items, item.Clone())
	}
	return items
}

func (v *BKV) ContainsKey(key interface{}) bool {
	for _, item := range v.kvs {
		switch k := key.(type) {
		case uint64:
			if item.NumberKey() == k {
				return true
			}

		case string:
			if item.StringKey() == k {
				return true
			}
		}
	}

	return false
}

func (v *BKV) Dump() {
	for _, item := range v.Items() {
		fmt.Printf("kv: %v -> %v", item.Key(), hex.EncodeToString(item.Value()))
	}
}

func (v *BKV) AddKV(kv *KV) {
	v.kvs = append(v.kvs, kv)
}


func (v *BKV) AddByUInt64Key(key uint64, value []byte) {
	kv := NewKVFromIntegerKey(key, value)
	v.AddKV(kv)
}

func (v *BKV) AddByStringKey(key string, value []byte) {
	kv := NewKVFromStringKey(key, value)
	v.AddKV(kv)
}

func (v *BKV) GetKVByStringKey(key string) *KV {
	for _, item := range v.kvs {
		if item.StringKey() == key {
			return item
		}
	}
	return nil
}

func (v *BKV) GetKVByNumberKey(key uint64) *KV {
	for _, item := range v.kvs {
		if item.NumberKey() == key {
			return item
		}
	}
	return nil
}

// by string key

func (v *BKV) GetRequiredNumberValueByStringKey(key string) (uint64, error) {
	kv := v.GetKVByStringKey(key)
	if kv == nil {
		return 0, ErrKVNotExists
	}

	return kv.NumberValue(), nil
}

func (v *BKV) GetNumberValueByStringKey(key string, defaultValue uint64) uint64 {
	kv := v.GetKVByStringKey(key)
	if kv == nil {
		return defaultValue
	}

	return kv.NumberValue()
}

func (v *BKV) GetRequiredStringValueByStringKey(key string) (string, error) {
	kv := v.GetKVByStringKey(key)
	if kv == nil {
		return "", ErrKVNotExists
	}

	return kv.StringValue(), nil
}

func (v *BKV) GetStringValueByStringKey(key string, defaultValue string) string {
	kv := v.GetKVByStringKey(key)
	if kv == nil {
		return defaultValue
	}

	return kv.StringValue()
}

// by number key

func (v *BKV) GetRequiredNumberValueByNumberKey(key uint64) (uint64, error) {
	kv := v.GetKVByNumberKey(key)
	if kv == nil {
		return 0, ErrKVNotExists
	}

	return kv.NumberValue(), nil
}

func (v *BKV) GetNumberValueByNumberKey(key uint64, defaultValue uint64) uint64 {
	kv := v.GetKVByNumberKey(key)
	if kv == nil {
		return defaultValue
	}

	return kv.NumberValue()
}

func (v *BKV) GetRequiredStringValueByNumberKey(key uint64) (string, error) {
	kv := v.GetKVByNumberKey(key)
	if kv == nil {
		return "", ErrKVNotExists
	}

	return kv.StringValue(), nil
}

func (v *BKV) GetStringValueByNumberKey(key uint64, defaultValue string) string {
	kv := v.GetKVByNumberKey(key)
	if kv == nil {
		return defaultValue
	}

	return kv.StringValue()
}

File: python/bkv.py

from typing import Union, Optional, List


def encode_length(i: int) -> bytearray:
    assert type(i) is int, "i must be an integer"

    if i < 0:
        raise RuntimeError('i must be a positive integer' % i)

    bs = bytearray()
    while i > 0:
        v = (i & 0x7F) | 0x80
        bs.append(v)
        i = i >> 7

    bs.reverse()
    bs[-1] &= 0x7F

    return bs


def decode_length(buf: bytearray):
    bs = bytearray()
    length_byte_size = 0
    for i, v in enumerate(buf):
        bs.append(v & 0x7F)
        length_byte_size += 1
        if v & 0x80 == 0:
            break

    if length_byte_size == 0 or len(bs) > 4:
        raise RuntimeError('invalid length buf: %s' % buf.hex())

    length = 0
    for _, v in enumerate(bs):
        length <<= 7
        length |= v

    return length, len(bs), buf[length_byte_size:]


def encode_number(i: int) -> bytearray:
    assert type(i) is int, "i must be an integer"

    if i < 0:
        raise RuntimeError('i must be a positive integer' % i)

    if i == 0:
        return bytearray()

    buf = bytearray()
    while i > 0:
        buf.append(i & 0xFF)
        i >>= 8
    buf.reverse()

    return buf


def decode_number(buf: bytearray) -> int:
    i = 0
    if len(buf) > 8:
        buf = buf[0:8]

    for _, v in enumerate(buf):
        i <<= 8
        i |= v

    return i


def encode_key(v: Union[str, int]) -> (bool, bytearray):
    if isinstance(v, str):
        return True, bytearray(v.encode())
    elif isinstance(v, int):
        return False, encode_number(v)
    else:
        raise RuntimeError('invalid key: {}'.format(v))


def encode_value(v: Union[str, int, bytearray]) -> bytearray:
    if isinstance(v, str):
        return bytearray(v.encode())
    elif isinstance(v, int):
        return encode_number(v)
    elif isinstance(v, bytearray):
        return v
    else:
        raise RuntimeError('invalid key: {}'.format(v))


class KV:
    is_string_key = False
    raw_key = bytearray()
    raw_value = bytearray()

    def __init__(self, k: Union[str, int] = None, v: Union[str, int, bytearray] = None):
        if k is not None:
            is_string_key, key = encode_key(k)
            self.is_string_key = is_string_key
            self.raw_key = key

        if v is not None:
            self.raw_value = encode_value(v)

    def pack(self) -> bytearray:
        buf = bytearray()

        key_len = len(self.raw_key)
        if key_len > 128:
            raise RuntimeError('key is too long(%d)' % key_len)

        total_len = 1 + key_len + len(self.raw_value)
        key_len_byte = key_len & 0x7F
        if self.is_string_key:
            key_len_byte |= 0x80

        buf += encode_length(total_len)
        buf.append(key_len_byte)
        buf += self.raw_key
        buf += self.raw_value

        return buf

    @classmethod
    def unpack(cls, bs: bytearray):
        if len(bs) == 0:
            return None, None

        total_len, _, pb = decode_length(bs)
        if len(pb) < total_len or len(pb) == 0:
            raise RuntimeError('total_len=%d, len(pb)=%d' % (total_len, len(pb)))

        key_len_byte = pb[0]
        key_len = key_len_byte & 0x7F
        is_string_key = False
        if key_len_byte & 0x80 != 0:
            is_string_key = True

        if key_len + 1 > total_len:
            raise RuntimeError('key_len bigger than total_len, key_len=%d, total_len=%d' % (key_len, total_len))

        key = pb[1:(key_len + 1)]
        value = pb[(key_len + 1):total_len]
        pending_bytes = pb[total_len:]

        kv = KV()
        kv.is_string_key = is_string_key
        kv.raw_key = key
        kv.raw_value = value
        return kv, pending_bytes

    def string_key(self):
        return self.raw_key.decode()

    def number_key(self):
        return decode_number(self.raw_key)

    def key(self) -> Union[str, int]:
        if self.is_string_key:
            return self.string_key()
        else:
            return self.number_key()

    def string_value(self):
        return self.raw_value.decode()

    def number_value(self):
        return decode_number(self.raw_value)


class BKV:
    kvs: List[KV] = []

    def pack(self) -> bytearray:
        buf = bytearray()
        for _, kv in enumerate(self.kvs):
            buf += kv.pack()

        return buf

    @classmethod
    def unpack(cls, bs: bytearray):
        kvs: List[KV] = []
        while True:
            if len(bs) == 0:
                break
            kv, pending_bytes = KV.unpack(bs)
            kvs.append(kv)
            bs = pending_bytes
        bkv = BKV()
        bkv.kvs = kvs
        return bkv

    def __getitem__(self, i: int):
        return self.kvs[i]

    def __setitem__(self, i: int, value: KV):
        self.kvs[i] = value

    def __len__(self):
        return len(self.kvs)

    def add(self, kv: KV):
        self.kvs.append(kv)
